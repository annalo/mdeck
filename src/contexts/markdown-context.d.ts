declare namespace MarkdownContext {
  declare interface ProviderProps {
    children: React.ReactNode;
  }

  declare interface State {
    htmlString: string;
    md: string;
    slideshowLineNumber: number;
    textLineNumber: number;
  }

  const enum ReducerActionType {
    SetHtmlString = "SET_HTML_STRING",
    SetMd = "SET_MD",
    SetSlideshowLineNumber = "SET_SLIDESHOW_LINE_NUMBER",
    SetTextLineNumber = "SET_TEXT_LINE_NUMBER",
  }

  declare type ReducerAction =
    | {
        type: ReducerActionType.SetHtmlString;
        htmlString: string;
      }
    | { type: ReducerActionType.SetMd; md: string }
    | {
        type: ReducerActionType.SetSlideshowLineNumber;
        slideshowLineNumber: number;
      }
    | {
        type: ReducerActionType.SetTextLineNumber;
        textLineNumber: number;
      };
}
