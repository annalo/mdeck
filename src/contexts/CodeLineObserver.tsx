import React, {
  createContext,
  useCallback,
  useContext,
  useMemo,
  useState,
} from "react";
import * as R from "ramda";

const CODE_LINE_OBSERVER_DEFAULT_INITIAL_STATE = {};

const CodeLineEntriesContext = createContext<
  CodeLineObserver.Entries | undefined
>(undefined);
const CodeLineObserverContext = createContext<
  CodeLineObserver.Observer | undefined
>(undefined);

const CodeLineObserverProvider: React.FC<CodeLineObserver.ProviderProps> = ({
  children,
  initialEntries,
}: CodeLineObserver.ProviderProps) => {
  const [entries, setEntries] = useState(
    initialEntries || CODE_LINE_OBSERVER_DEFAULT_INITIAL_STATE
  );
  const observe: CodeLineObserver.Observe = useCallback(
    (targetEntries) =>
      setEntries((currentEntries) =>
        R.mergeRight(currentEntries, targetEntries)
      ),
    []
  );

  const observerProviderValue = useMemo(() => ({ observe }), [observe]);
  return (
    <CodeLineEntriesContext.Provider value={entries}>
      <CodeLineObserverContext.Provider value={observerProviderValue}>
        {children}
      </CodeLineObserverContext.Provider>
    </CodeLineEntriesContext.Provider>
  );
};

function useCodeLineEntries(): CodeLineObserver.Entries {
  const context = useContext(CodeLineEntriesContext);
  if (context === undefined) {
    throw new Error(
      "useCodeLineEntries must be used within CodeLineObserverProvider or SlideshowObservable"
    );
  }
  return context;
}

function useCodeLineObserver(): CodeLineObserver.Observer {
  const context = useContext(CodeLineObserverContext);
  if (context === undefined) {
    throw new Error(
      "useCodeLineObserver must be used within CodeLineObserverProvider or SlideshowObservable"
    );
  }
  return context;
}

export { CodeLineObserverProvider, useCodeLineEntries, useCodeLineObserver };
