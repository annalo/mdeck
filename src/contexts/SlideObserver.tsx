import React, {
  createContext,
  useCallback,
  useContext,
  useMemo,
  useState,
} from "react";
import * as R from "ramda";

const SLIDE_ENTRIES_INITIAL_STATE = {};

const SlideEntriesContext = createContext<SlideObserver.Entries | undefined>(
  undefined
);
const SlideObserverContext = createContext<SlideObserver.Observer | undefined>(
  undefined
);

const SlideObserverProvider: React.FC<SlideObserver.ProviderProps> = ({
  children,
  initialEntries,
}: SlideObserver.ProviderProps) => {
  const [entries, setEntries] = useState<SlideObserver.Entries>(
    initialEntries || SLIDE_ENTRIES_INITIAL_STATE
  );
  const observe: SlideObserver.Observe = useCallback(
    (slideNumber, targetElement) =>
      setEntries((currentEntries) =>
        R.mergeRight(currentEntries, { [slideNumber]: targetElement })
      ),
    []
  );
  const unobserve: SlideObserver.Unobserve = useCallback(
    (slideNumber: SlideNumber) =>
      setEntries((currentEntries) => R.omit([slideNumber], currentEntries)),
    []
  );

  const observerProviderValue = useMemo(() => ({ observe, unobserve }), [
    observe,
    unobserve,
  ]);
  return (
    <SlideEntriesContext.Provider value={entries}>
      <SlideObserverContext.Provider value={observerProviderValue}>
        {children}
      </SlideObserverContext.Provider>
    </SlideEntriesContext.Provider>
  );
};

function useSlideEntries(): SlideObserver.Entries {
  const context = useContext(SlideEntriesContext);
  if (context === undefined) {
    throw new Error(
      "useSlideEntries must be used within a SlideObserverProvider"
    );
  }
  return context;
}

function useSlideObserver(): SlideObserver.Observer {
  const context = useContext(SlideObserverContext);
  if (context === undefined) {
    throw new Error(
      "useSlideObserver must be used within a SlideObserverProvider"
    );
  }
  return context;
}

export { SlideObserverProvider, useSlideEntries, useSlideObserver };
