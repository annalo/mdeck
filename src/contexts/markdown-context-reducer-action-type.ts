export const enum MarkdownContextReducerActionType {
  SetHtmlString = "SET_HTML_STRING",
  SetMd = "SET_MD",
  SetSlideshowLineNumber = "SET_SLIDESHOW_LINE_NUMBER",
  SetTextLineNumber = "SET_TEXT_LINE_NUMBER",
}
