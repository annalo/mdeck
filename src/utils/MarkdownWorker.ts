import Worker from "worker-loader!./Worker"; // eslint-disable-line
import type { Dispatch } from "react";
import type { MarkdownContextReducerAction } from "types/markdown-context";

export class MarkdownWorker {
  private readonly worker: Worker;

  constructor(dispatch: Dispatch<MarkdownContextReducerAction>) {
    this.worker = new Worker();
    this.worker.onmessage = ({
      data,
    }: {
      data: MarkdownContextReducerAction;
    }) => {
      console.log("markdown worker on message", data);
      dispatch(data);
    };
  }

  parse(md: string): void {
    this.worker.postMessage({ action: "PARSE", md });
  }

  terminate() {
    if (this.worker) this.worker.terminate();
  }
}
