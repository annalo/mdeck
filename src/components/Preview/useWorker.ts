import { useEffect, useRef } from "react";
import type { Dispatch } from "react";
import type { MarkdownContextReducerAction } from "types/markdown-context-reducer-action";

import MarkdownWorker from "./markdown-worker";

interface UseWorkerProps {
  dispatch: Dispatch<MarkdownContextReducerAction>;
  md: MarkdownString;
}

function useWorker({ dispatch, md }: UseWorkerProps): void {
  const workerRef = useRef<MarkdownWorker | null>(null);

  useEffect(() => {
    workerRef.current = new MarkdownWorker(dispatch);
    return () => {
      workerRef.current?.terminate();
    };
  }, [dispatch]);

  useEffect(() => {
    const worker = workerRef.current;
    worker?.parse(md);
  }, [md]);
}

export { useWorker };
