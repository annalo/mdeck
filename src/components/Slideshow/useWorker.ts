import { useEffect, useRef } from "react";
import type { Dispatch } from "react";
import type { MarkdownContextReducerAction } from "types/markdown-context";

import { MarkdownWorker } from "utils/MarkdownWorker";

interface UseWorkerProps {
  dispatch: Dispatch<MarkdownContextReducerAction>;
  md: MarkdownString;
}

export const useWorker = ({ dispatch, md }: UseWorkerProps): void => {
  const workerRef = useRef<MarkdownWorker | null>(null);

  useEffect(() => {
    workerRef.current = new MarkdownWorker(dispatch);
    return () => {
      workerRef.current?.terminate();
    };
  }, [dispatch]);

  useEffect(() => {
    const worker = workerRef.current;
    worker?.parse(md);
  }, [md]);
};
