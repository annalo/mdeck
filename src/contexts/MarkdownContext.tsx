import React, { createContext, useReducer } from "react";
import { parse } from "utils/parse";

interface MarkdownProviderProps {
  children: React.ReactElement[];
}

interface MarkdownContextInitialState {
  activePane: string;
  html: string;
  previewLineNumber: number;
  textLineNumber: number;
  lineNumber: number;
  md: string;
}

const MARKDOWN_CONTEXT_INITIAL_STATE = {
  activePane: "",
  html: "",
  previewLineNumber: 0,
  textLineNumber: 0,
  lineNumber: 0,
  md: "",
};

export const MarkdownContext = createContext<{
  state: MarkdownContextInitialState;
  dispatch: React.Dispatch<any>;
}>({
  state: MARKDOWN_CONTEXT_INITIAL_STATE,
  dispatch: () => null,
});

export const MarkdownProvider: React.FC<MarkdownProviderProps> = ({
  children,
}: MarkdownProviderProps) => {
  function reducer(
    state,
    { type, activePane, textLineNumber, previewLineNumber, lineNumber, md }
  ) {
    switch (type) {
      case "setActivePane":
        return { ...state, activePane };
      case "setTextLineNumber":
        return { ...state, textLineNumber };
      case "setPreviewLineNumber":
        return { ...state, previewLineNumber };
      case "setLineNumber":
        return { ...state, lineNumber };
      case "setMd":
        return { ...state, html: parse(md), md };
      default:
        throw new Error();
    }
  }

  const [state, dispatch] = useReducer(reducer, MARKDOWN_CONTEXT_INITIAL_STATE);
  const context = { state, dispatch };

  return (
    <MarkdownContext.Provider value={context}>
      {children}
    </MarkdownContext.Provider>
  );
};
