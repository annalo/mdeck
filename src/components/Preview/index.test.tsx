import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";

import {
  MARKDOWN_CONTEXT_DEFAULT_INITIAL_STATE,
  MarkdownProvider,
  useMarkdownDispatch,
} from "contexts/MarkdownContext";
import { CodeLineObserverProvider } from "contexts/CodeLineObserver";
import { SlideObserverProvider } from "contexts/SlideObserver";

import { Preview } from ".";
import MarkdownWorker from "./markdown-worker";
import { useWorker } from "./useWorker";

jest.mock("screenfull");
jest.mock("./markdown-worker");

afterEach(() => jest.clearAllMocks());

describe("<Preview />", () => {
  const wrapper = ({ children }) => (
    <MarkdownProvider>
      <SlideObserverProvider>
        <CodeLineObserverProvider>{children}</CodeLineObserverProvider>
      </SlideObserverProvider>
    </MarkdownProvider>
  );
  test("should render and match the snapshot", () => {
    const { asFragment } = render(<Preview />, { wrapper });
    expect(asFragment()).toMatchSnapshot();
  });

  describe("useWorker", () => {
    test("should instantiate MarkdownWorker just once", () => {
      let md = "## Markdown String";
      const { rerender } = renderHook(
        () => {
          const dispatch = useMarkdownDispatch();
          useWorker({ dispatch, md });
        },
        { wrapper }
      );

      md = "## Markdown String\n* Bullet 1";
      rerender();

      expect(MarkdownWorker).toHaveBeenCalledTimes(1);
    });

    test("should parse md with MarkdownWorker", () => {
      let md = "## Markdown String";
      const { rerender } = renderHook(
        () => {
          const dispatch = useMarkdownDispatch();
          useWorker({ dispatch, md });
        },
        { wrapper }
      );

      md = "## Markdown String\n* Bullet 1";
      rerender();

      const workerInstance = MarkdownWorker.mock.instances[0];
      expect(workerInstance.parse).toHaveBeenCalledTimes(2);
    });
  });

  describe("usePresentation", () => {
    const htmlArray = [
      "<svg data-line='1'>1</svg>",
      "<svg data-line='2'>2</svg>",
      "<svg data-line='3'>3</svg>",
    ];
    const wrapperWithMarkdownValues = ({ children }) => (
      <MarkdownProvider
        initialValues={{ ...MARKDOWN_CONTEXT_DEFAULT_INITIAL_STATE, htmlArray }}
      >
        <SlideObserverProvider>
          <CodeLineObserverProvider>{children}</CodeLineObserverProvider>
        </SlideObserverProvider>
      </MarkdownProvider>
    );

    test("should not be in fullscreen by default", () => {
      render(<Preview />, { wrapper: wrapperWithMarkdownValues });
      expect(document.fullscreen).not.toBeTruthy();
    });

    test("should fullscreen slideshow when fullscreen button is clicked", async () => {
      render(<Preview />, { wrapper: wrapperWithMarkdownValues });

      expect(document.fullscreen).not.toBeTruthy();

      const button = screen.getByRole("button");
      fireEvent.click(button);
      waitFor(() => expect(document.fullscreen).toBeTruthy());
    });
  });
});
