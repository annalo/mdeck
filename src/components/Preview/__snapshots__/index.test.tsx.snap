// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Preview /> should render and match the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
    class="c0"
    id="preview"
  />
</DocumentFragment>
`;
